internal/provider/datasources/deployment.go

```diff
@@ -0,0 +1,405 @@
+package datasources
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/google/uuid"
+	"github.com/hashicorp/terraform-plugin-framework-jsontypes/jsontypes"
+	"github.com/hashicorp/terraform-plugin-framework/datasource"
+	"github.com/hashicorp/terraform-plugin-framework/datasource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/path"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/prefecthq/terraform-provider-prefect/internal/api"
+	"github.com/prefecthq/terraform-provider-prefect/internal/provider/customtypes"
+	"github.com/prefecthq/terraform-provider-prefect/internal/provider/helpers"
+	"github.com/prefecthq/terraform-provider-prefect/internal/provider/resources"
+)
+
+// Ensure the implementation satisfies the expected interfaces.
+var (
+	_ datasource.DataSource              = &deploymentDataSource{}
+	_ datasource.DataSourceWithConfigure = &deploymentDataSource{}
+)
+
+// deploymentDataSource is the data source implementation.
+type deploymentDataSource struct {
+	client api.PrefectClient
+}
+
+// DeploymentDataSourceModel defines the Terraform data source model.
+type DeploymentDataSourceModel struct {
+	// The model requires the same fields, so reuse the fields defined in the resource model.
+	resources.DeploymentResourceModel
+}
+
+// NewDeploymentDataSource is a helper function to simplify the provider implementation.
+//
+//nolint:ireturn // required by Terraform API
+func NewDeploymentDataSource() datasource.DataSource {
+	return &deploymentDataSource{}
+}
+
+// Metadata returns the data source type name.
+func (d *deploymentDataSource) Metadata(_ context.Context, req datasource.MetadataRequest, resp *datasource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_deployment"
+}
+
+// Schema defines the scema for the data source.
+func (d *deploymentDataSource) Schema(_ context.Context, _ datasource.SchemaRequest, resp *datasource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		Description: `
+Get information about an existing Deployment by either:
+- deployment ID, or
+- deployment name
+The Deployment ID takes precedence over deployment name.
+`,
+		Attributes: map[string]schema.Attribute{
```

Comment:
I'd love to just copy the attributes from the Deployment resource, but the fields on each attribute are a little different for data sources.